### ISSUE TITLE
Stuttering in Gundam VS Gundam NEXTPLUS

### ISSUE THREAD

**Issue by daniel229** [hrydgard/ppsspp#8583]
Since https://github.com/hrydgard/ppsspp/pull/8292

**Comment by unknownbrackets** [comment]
Does it stutter less often if you change `#define VERTEXCACHE_NAME_DECIMATION_INTERVAL 41` to like 410?

If the vertex cache is creating a huge number of buffers that might be its own kind of problem...

-[Unknown]

**Comment by daniel229** [comment]
Look like still the same even 4100,41000

**Comment by unknownbrackets** [comment]
Hmm.  If you go back to 41, what if you add before this:

```c++
if (bufferNameCacheSize_ > VERTEXCACHE_NAME_CACHE_FULL_BYTES) {
```

This:

```c++
NOTICE_LOG(HLE, "Buffers: %lld bytes among %lld", (u64)bufferNameCacheSize_, (u64)bufferNameCache_.size());
```

Also, you have GfxDebugOutput off, right?

-[Unknown]

**Comment by daniel229** [comment]
Yes, GfxDebugOutput is off,I don't see it in the ppsspp.ini.

log
https://gist.github.com/daniel229/35832785ec4ea875c015

**Comment by unknownbrackets** [comment]
Buffers: 7717210 bytes among 6155

Hmm, quite a few buffers.  About 1KB each.

Hmm, what do you think @hrydgard?  Maybe it needs a better allocation strategy in AllocateBuffer?  Might just be stuttering due to the time to look up a buffer of the right size, I guess.

-[Unknown]

**Comment by hrydgard** [comment]
If you turn off vertex cache, what's the performance like?

VAOs are poorly implemented in many OpenGL graphics drivers due to low usage in practice. I'm starting to think we should ditch them again, even though they are technically required for 3.0 core profile.

**Comment by daniel229** [comment]
Stuttering is gone with vertex cache off.
In this map the vertex cache off is a little faster than it on.
vertex cache off
![off1](https://cloud.githubusercontent.com/assets/3481559/13027270/04df41ac-d283-11e5-872f-88ce488cdd23.png)

vertex cache on
![on1](https://cloud.githubusercontent.com/assets/3481559/13027269/04dd3f06-d283-11e5-900d-c74ec69ca662.png)

In the slowest map,vertex cache off is 33%+ slower than it on.
vertex cache off
![off2](https://cloud.githubusercontent.com/assets/3481559/13027304/68e67ff2-d284-11e5-8804-6ed82d05aadd.png)

vertex cache on
![on2](https://cloud.githubusercontent.com/assets/3481559/13027305/68ec7862-d284-11e5-897f-732b1b91f04d.png)

**Comment by unknownbrackets** [comment]
> VAOs are poorly implemented in many OpenGL graphics drivers due to low usage in practice. I'm starting to think we should ditch them again, even though they are technically required for 3.0 core profile.

Note that we are not using them for everything when the core flag is off, which is the case here.  This is just for the vertex cache buffers.

-[Unknown]

**Comment by hrydgard** [comment]
Ah, right. But something changed in #8292 .. but maybe new buffer name behaviour has some bad case?

Either way, just play this game with no vertex cache for now... 

All this stuff will be so much better in Vulkan.

**Comment by unknownbrackets** [comment]
I'm thinking maybe using a `std::multimap` for a lookup by size or something might be faster for allocation.  Or in general, if the list was kept sorted by size, it could start in the middle or something.

-[Unknown]

**Comment by daniel229** [comment]
Also affect FF2

**Comment by hrydgard** [comment]
I think the remaining perf problems can be closed as a duplicate of #1319.
